kakao/KakaoLoginServer/app.js

```diff
@@ -33,47 +36,112 @@ function requestMe(kakaoAccessToken) {
+ * @param  {String} kakaoUserId    user id per app
+ * @param  {String} email          user's email address
+ * @param  {Boolean} emailVerified whether this email is verified or not
+ * @param  {String} displayName    user
+ * @param  {String} photoURL       profile photo url
+ * @return {Promise<UserRecord>}   Firebase user record in a promise
  */
-function updateOrCreateUser(userId, email, displayName, photoURL) {
-  console.log('updating or creating a firebase user');
-  const updateParams = {
-    provider: 'KAKAO',
-    displayName: displayName,
-  };
-  if (displayName) {
-    updateParams['displayName'] = displayName;
-  } else {
-    updateParams['displayName'] = email;
-  }
-  if (photoURL) {
-    updateParams['photoURL'] = photoURL;
-  }
-  console.log(updateParams);
-  return firebaseAdmin.auth().updateUser(userId, updateParams)
-  .catch((error) => {
-    if (error.code === 'auth/user-not-found') {
-      updateParams['uid'] = userId;
-      if (email) {
-        updateParams['email'] = email;
+function createOrLinkUser(kakaoUserId, email, emailVerified, displayName,
+                            photoURL) {
+  return getUser(kakaoUserId, email, emailVerified)
+    .catch((error) => {
+      if (error.code === 'auth/user-not-found') {
+        const params = {
+          uid: `kakao:${kakaoUserId}`,
+          displayName: displayName,
+        };
+        if (email) {
+          params['email'] = email;
+        }
+        if (photoURL) {
+          params['photoURL'] = photoURL;
+        }
+        console.log(`creating a firebase user with email ${email}`);
+        return firebaseAdmin.auth().createUser(params);
       }
-      return firebaseAdmin.auth().createUser(updateParams);
-    }
-    throw error;
-  });
-};
+      throw error;
+    })
+    .then((userRecord) => linkUserWithKakao(kakaoUserId, userRecord));
+}
 
+/**
+ * getUser - fetch firebase user with kakao UID first, then with email if
+ * no user found. If email is not verified, throw an error so that
+ * the user can re-authenticate.
+ *
+ * @param {String} kakaoUserId    user id per app
+ * @param {String} email          user's email address
+ * @param {Boolean} emailVerified whether this email is verified or not
+ * @return {Promise<admin.auth.UserRecord>}
+ */
+function getUser(kakaoUserId, email, emailVerified) {
+  console.log(`fetching a firebase user with uid kakao:${kakaoUserId}`);
+  return firebaseAdmin.auth().getUser(`kakao:${kakaoUserId}`)
+    .catch((error) => {
+      if (error.code !== 'auth/user-not-found') {
+        throw error;
+      }
+      if (!email) {
+        throw error; // cannot find existing accounts since there is no email.
+      }
+      console.log(`fetching a firebase user with email ${email}`);
+      return firebaseAdmin.auth().getUserByEmail(email)
+        .then((userRecord) => {
+          if (!emailVerified) {
+            throw new Error('This user should authenticate first ' +
```

Comment:
Here I would return a special error with a specific code so that we can act a special way when this happens. 
Currently a new User gets created. Maybe do:

```js
const error =  new Error('This user should authenticate first ' +
              'with other providers');
error.code = 'auth/existing-account-need-auth';
error.lineAccessToken = kakaoAccessToken; // Pass the access Token up to here, we'll need it for later.
throw error;
```